module/apmslog/handler_test.go

```diff
@@ -0,0 +1,147 @@
+	var buf bytes.Buffer
+
+	h := newApmslogHandler(&buf, tracer)
+	logger := slog.New(h)
+
+	logger.Error("hello world", "error", errors.New("new error"))
+
+	assert.Equal(t, `{"time":"1970-01-01T00:00:00Z","level":"ERROR","msg":"hello world","error":"new error"}`+"\n", buf.String())
+
+	tracer.Flush(nil)
+	payloads := transport.Payloads()
+	assert.Len(t, payloads.Errors, 1)
+
+	err0 := payloads.Errors[0]
+	assert.Equal(t, "hello world", err0.Log.Message)
+	assert.Equal(t, "error", err0.Log.Level)
+	assert.Equal(t, "", err0.Log.LoggerName)
+	assert.Equal(t, "", err0.Log.ParamMessage)
+	assert.Equal(t, "TestHandler", err0.Culprit)
+	assert.NotEmpty(t, err0.Log.Stacktrace)
+	// assert.Equal(t, model.Time(time.Unix(0, 0).UTC()), err0.Timestamp) // seems like slog time attribute is not accessable
+	assert.Zero(t, err0.ParentID)
+	assert.Zero(t, err0.TraceID)
+	assert.Zero(t, err0.TransactionID)
+}
+func TestHandlerTransactionTraceContext(t *testing.T) {
+	tracer, transport := transporttest.NewRecorderTracer()
+	defer tracer.Close()
+
+	var buf bytes.Buffer
+
+	h := newApmslogHandler(&buf, tracer)
+	logger := slog.New(h)
+
+	tx := tracer.StartTransaction("name", "type")
+	ctx := apm.ContextWithTransaction(context.Background(), tx)
+	span, ctx := apm.StartSpan(ctx, "name", "type")
+
+	logger.ErrorContext(ctx, "hello world", "error", errors.New("new error"))
+
+	span.End()
+	tx.End()
+
+	tracer.Flush(nil)
+	payloads := transport.Payloads()
+	assert.Len(t, payloads.Transactions, 1)
+	assert.Len(t, payloads.Spans, 1)
+	assert.Len(t, payloads.Errors, 1)
+
+	err0 := payloads.Errors[0]
+	assert.Equal(t, payloads.Spans[0].ID, err0.ParentID)
+	assert.Equal(t, payloads.Transactions[0].TraceID, err0.TraceID)
+	assert.Equal(t, payloads.Transactions[0].ID, err0.TransactionID)
+}
+
+func TestHandlerWithError(t *testing.T) {
+	tracer, transport := transporttest.NewRecorderTracer()
+	defer tracer.Close()
+
+	var buf bytes.Buffer
+
+	h := newApmslogHandler(&buf, tracer)
+	logger := slog.New(h)
+
+	logger.Error("hello world", "error", mockFuncError())
+
+	tracer.Flush(nil)
+	payloads := transport.Payloads()
+	assert.Len(t, payloads.Errors, 1)
+
+	err0 := payloads.Errors[0]
+	assert.Equal(t, "new error", err0.Exception.Message)
+	assert.Equal(t, "hello world", err0.Log.Message)
+	assert.Equal(t, "mockFuncError", err0.Culprit)
+	assert.NotEmpty(t, err0.Log.Stacktrace)
+	assert.NotEmpty(t, err0.Exception.Stacktrace)
+	assert.NotEqual(t, err0.Log.Stacktrace, err0.Exception.Stacktrace)
+	assert.Equal(t, "mockFuncError", err0.Exception.Stacktrace[0].Function)
+	assert.Equal(t, "(*Logger).log", err0.Log.Stacktrace[0].Function)
+}
+
+func mockFuncError() error {
+	return errors.New("new error")
+}
+
+func TestHandlerTracerClosed(t *testing.T) {
+	tracer, _ := transporttest.NewRecorderTracer()
+	tracer.Close() // close it straight away, hook should return immediately
+
+	h := newApmslogHandler(io.Discard, tracer)
+	logger := slog.New(h)
+	logger.Error("fail")
```

Comment:
No expectation here? Wouldn't the test pass exactly the same way if we didn't call `tracer.Close()`?

Reply:
this is to test that the logger still works as expected (as a logger) even if the tracer closes. I will add an assert.